src/abstractdatatable/join.jl

```diff
@@ -214,6 +144,170 @@ function sharepools(dt1::AbstractDataTable, dt2::AbstractDataTable)
     sharepools(refs1, refs2)
 end
 
+# helper structure for DataTables joining
+immutable DataTableJoiner{DT1<:AbstractDataTable, DT2<:AbstractDataTable}
+    dtl::DT1
+    dtr::DT2
+    dtl_on::DT1
+    dtr_on::DT2
+    on_cols::Vector{Symbol}
+
+    function DataTableJoiner(dtl::DT1, dtr::DT2, on::Union{Symbol,Vector{Symbol}})
+        on_cols = (isa(on, Symbol) ? fill(on, 1) : on)
+        new(dtl, dtr, dtl[on_cols], dtr[on_cols], on_cols)
+    end
+end
+
+DataTableJoiner{DT1<:AbstractDataTable, DT2<:AbstractDataTable}(dtl::DT1, dtr::DT2, on::Union{Symbol,Vector{Symbol}}) =
+    DataTableJoiner{DT1,DT2}(dtl, dtr, on)
+
+# helper map between the row indices in original and joined table
+immutable RowIndexMap
+    "row indices in the original table"
+    orig::Vector{Int}
+    "row indices in the resulting joined table"
+    join::Vector{Int}
+end
+
+Base.length(x::RowIndexMap) = length(x.orig)
+
+# fix the result of the rightjoin by taking the nonnull values from the right table
+function fix_rightjoin_column!(res_col::AbstractArray, col_ix::Int, joiner::DataTableJoiner,
+                               all_orig_left_ixs::Vector{Int}, rightonly_ixs::RowIndexMap)
+    res_col[rightonly_ixs.join] = joiner.dtr_on[rightonly_ixs.orig, col_ix]
+    res_col
+end
+
+# composes the joined data table using the maps between the left and right
+# table rows and the indices of rows in the result
+function compose_joined_table(joiner::DataTableJoiner,
+                              left_ixs::RowIndexMap, leftonly_ixs::RowIndexMap,
+                              right_ixs::RowIndexMap, rightonly_ixs::RowIndexMap)
+    @assert length(left_ixs) == length(right_ixs)
+    # compose left half of the result taking all left columns
+    all_orig_left_ixs = vcat(left_ixs.orig, leftonly_ixs.orig)
+    if length(leftonly_ixs) > 0
+        # permute the indices to restore left table rows order
+        all_orig_left_ixs[vcat(left_ixs.join, leftonly_ixs.join)] = all_orig_left_ixs
```

Comment:
Woops, that won't work because the RHS and LHS are the same vector. Calling `copy(all_orig_left_ixs)` and using it for the RHS should do the trick. It's what `setindex!` does under the hood, but we can avoid another copy by not calling `vcat`.

Also add `@inbounds`, which is another advantage of the custom solution (since we know that indices are correct).